File: week-2/group-of-special-equivalent-strings/group-of-special-equivalent-strings.py
class Solution:
    def numSpecialEquivGroups(self, A: List[str]) -> int:
        hashtable = {}
        for string in A:
            odd = [0]*26
            even = [0]*26
            for i in range(len(string)):
                if i%2 == 0:
                    even[ord(string[i]) - ord('a')] += 1
                else:
                    odd[ord(string[i]) - ord('a')] += 1
            state = str(odd) + str(even)
            hashtable[state] = True
        
        return len(hashtable.keys())
        
        

File: week-2/group-of-special-equivalent-strings/group-of-special-equivalent-strings.js
/**
 * @param {string[]} A
 * @return {number}
 */
var numSpecialEquivGroups = function(A) {
    const hashtable = {};
    for (let string of A) {
        const odd = new Array(26).fill(0);
        const even = new Array(26).fill(0);
        for (let i=0; i<string.length; i++) {
            if (i%2 === 0)
                ++even[string[i].charCodeAt() - 'a'.charCodeAt()];
            else
                ++odd[string[i].charCodeAt() - 'a'.charCodeAt()];
        }
        state = odd + '~' + even;
        hashtable[state] = true;
    }
    
    return Object.keys(hashtable).length;
};
